### PR TITLE
Object-tracking - Estimator

### PR DESCRIPTION
**Pull request by ss-sinha**
# Peer Review Information

Adding Estimator for SiamMask
PR in companion to #37 
# Pull Request Check List

- [ ] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

### CODE REVIEW

File: datasetinsights/estimators/siam_mask.py [hunk 1]
```diff
@@ -0,0 +1,2215 @@
...
+
+from .base import Estimator
+
+# from datasetinsights.evaluation_metrics.base import EvaluationMetric
+
+
+logger = logging.getLogger(__name__)
+
+
+def corner2center(corner):
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:34]: Should have docstring explanation of what is done.

File: datasetinsights/estimators/siam_mask.py [hunk 2]
```diff
@@ -0,0 +1,2215 @@
...
+    x1, y1, x2, y2 = corner[0], corner[1], corner[2], corner[3]
+    x = (x1 + x2) * 0.5
+    y = (y1 + y2) * 0.5
+    w = x2 - x1
+    h = y2 - y1
+    return x, y, w, h
+
+
+def center2corner(center):
+    """
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:52]: Docstring explanation.

File: datasetinsights/estimators/siam_mask.py (hunk 1 above)
Comment (zookae) [reply datasetinsights/estimators/siam_mask.py:34]: Also the formatting should match the canonical docstring format in other estimators. Roughly like:

"<description>

Args:
<indented list of args>

Returns:
<indented description of return value>

File: datasetinsights/estimators/siam_mask.py [hunk 3]
```diff
@@ -0,0 +1,2215 @@
...
+    Converting coordinates to standard BBox implementation
+    """
+    imh, imw = image.shape[:2]
+    if len(shape) == 4:
+        w, h = shape[2] - shape[0], shape[3] - shape[1]
+    else:
+        w, h = shape
+    context_amount = 0.5
+    exemplar_size = template_size  # 127
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]:
```suggestion
    exemplar_size = template_size
```

File: datasetinsights/estimators/siam_mask.py [hunk 4]
```diff
@@ -0,0 +1,2215 @@
...
+def toBBox(image, shape, template_size=127):
+    """
+    Converting coordinates to standard BBox implementation
+    """
+    imh, imw = image.shape[:2]
+    if len(shape) == 4:
+        w, h = shape[2] - shape[0], shape[3] - shape[1]
+    else:
+        w, h = shape
+    context_amount = 0.5
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:92]: what is `context_amount`?

Comment (zookae) [reply datasetinsights/estimators/siam_mask.py:92]: (and should it be a model config or input to `toBBox`?)

File: datasetinsights/estimators/siam_mask.py [hunk 5]
```diff
@@ -0,0 +1,2215 @@
...
+    Args:
+        pretrained (bool): If True, returns a model pre-trained on ImageNet
+    """
+    model = ResNet(Bottleneck, [3, 4, 6, 3], **kwargs)
+    # if pretrained:
+    #     model.load_state_dict(
+    #         model_zoo.load_url(
+    #             "https://download.pytorch.org/models/resnet50-19c8e357.pth"
+    #         )
+    #     )
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:2031]:
```suggestion
```

File: datasetinsights/estimators/siam_mask.py [hunk 6]
```diff
@@ -0,0 +1,2215 @@
...
+
+
+# Siamese network (shared ResNet)+(unshared ResDownS)
+class ResDown(MultiStageFeature):
+    def __init__(self, pretrain=False):
+        super(ResDown, self).__init__()
+        # features is the original Siamese network
+        # -> till layer3 of resnet 50 (shared)
+        self.features = resnet50(layer3=True, layer4=False)
+        # Weights for this need to be downloaded from the oxford link (TODO)
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]: is this still TODO?

File: datasetinsights/estimators/siam_mask.py [hunk 7]
```diff
@@ -0,0 +1,2215 @@
...
+                )
+                torch.nn.utils.clip_grad_norm_(
+                    model.module.rpn_model.parameters(), config["clip"]["rpn"]
+                )
+                torch.nn.utils.clip_grad_norm_(
+                    model.module.mask_model.parameters(), config["clip"]["mask"]
+                )
+
+            def is_valid_number(x):
+                return not (math.isnan(x) or math.isinf(x) or x > 1e4)
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]: why is this being declared inside the function?

File: datasetinsights/estimators/siam_mask.py [hunk 8]
```diff
@@ -0,0 +1,2215 @@
...
+
+            if is_valid_number(loss.item()):
+                optimizer.step()
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]: when would the loss be invalid? it seems odd to silently fail and keep going.

File: datasetinsights/estimators/siam_mask.py [hunk 9]
```diff
@@ -0,0 +1,2215 @@
...
+    cv2.imwrite(name, image)
+
+
+"""
+    Generates a random number between 2 given numbers
+"""
+
+
+def genarate_random_number(a, b):
+    return a + (random.random() * (b - a))
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:153]: seems mis-formatted

File: datasetinsights/estimators/siam_mask.py [hunk 10]
```diff
@@ -0,0 +1,2215 @@
...
+        (out_sz, out_sz),
+        borderMode=cv2.BORDER_CONSTANT,
+        borderValue=padding,
+    )
+    return crop
+
+
+def draw(image, box, name):
+    """
+    Only for debugging - will draw a bbox on image
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]: docstring

File: datasetinsights/estimators/siam_mask.py [hunk 11]
```diff
@@ -0,0 +1,2215 @@
...
+    h = h * scale_z
+    cx, cy = imw // 2, imh // 2
+    bbox = anyform2canonicalBBox([cx, cy, w, h], "center")
+    return bbox
+
+
+def crop_hwc(image, bbox, out_sz, padding=(0, 0, 0)):
+    """
+    Cropping image given a bounding box
+    """
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]: docstring

File: datasetinsights/estimators/siam_mask.py [hunk 12]
```diff
@@ -0,0 +1,2215 @@
...
+    "center2corner": center2corner,
+    "random_number": genarate_random_number,
+}
+
+
+def get_transforms(center_crop, blur, gray, shift, scale, resize, flip):
+
+    """
+    # Should maintain this order
+    # (x) TODO: center crop both template and search
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]: still TODO?

File: datasetinsights/estimators/siam_mask.py [hunk 13]
```diff
@@ -0,0 +1,2215 @@
...
+    if shift[0]:
+        transforms.append(ShiftBBox(helper=helper, shift=shift[1]))
+    if gray[0]:
+        transforms.append(GrayImage(p=gray[1]))
+    if blur[0]:
+        transforms.append(BlurImage(p=blur[1]))
+    # if resize[0]:
+    #     transforms.append(Resize())
+    # if flip[0]:
+    #     transforms.append(Flip())
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]: we should remove commented out code once it's being merged.

File: datasetinsights/estimators/siam_mask.py [hunk 14]
```diff
@@ -0,0 +1,2215 @@
...
+class SiamMaskCompose:
+    def __init__(self, transforms):
+        self.transforms = transforms
+        self.rgbVar = np.array(
+            [
+                [-0.55919361, 0.98062831, -0.41940627],
+                [1.72091413, 0.19879334, -1.82968581],
+                [4.64467907, 4.73710203, 4.88324118],
+            ],
+            dtype=np.float32,
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:210]: what are these magic numbers? should at least be explained/defined.

File: datasetinsights/estimators/siam_mask.py [hunk 15]
```diff
@@ -0,0 +1,2215 @@
...
+        )
+
+    def __call__(self, image, annotations, size, mask):
+
+        """
+        Applies the transforms populated in the transforms list
+        """
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:217]: docstring

File: datasetinsights/estimators/__init__.py [hunk 16]
```diff
@@ -3,4 +3,4 @@
 from .densedepth import DenseDepth
 from .faster_rcnn import FasterRCNN
 
-__all__ = [Estimator, DeeplabV3, DenseDepth, FasterRCNN]
+__all__ = [Estimator, DeeplabV3, DenseDepth, FasterRCNN]
```
Comment (zookae) [inline datasetinsights/estimators/__init__.py]: should the SiamMask estimator be exposed as well?

File: datasetinsights/estimators/siam_mask.py [hunk 17]
```diff
@@ -0,0 +1,2251 @@
...
+    cv2.rectangle(image, (x1, y1), (x2, y2), (0, 255, 0))
+    cv2.imwrite(name, image)
+
+
+"""
+    Generates a random number between 2 given numbers
+"""
+
+
+def genarate_random_number(a, b):
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:152]: Is this an integer sampled uniformly? if so, let's name this more clearly/explicitly.

File: datasetinsights/estimators/siam_mask.py [hunk 18]
```diff
@@ -0,0 +1,2251 @@
...
+        for t in self.transforms[3:]:
+            image, _ = t((image, None))
+
+        return image, bbox, mask, bbox_orig
+
+
+class Anchors:
+    def __init__(self, cfg):
+        """
+            Args: config file
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:283]: what config file?

File: datasetinsights/estimators/siam_mask.py [hunk 19]
```diff
@@ -0,0 +1,2251 @@
...
+        **kwargs,
+    ):
+
+        # config files should contain the hyperparameters of the model
+        # TB Writer object comes from the main class itself
+        # checkpointer should be used to save progress after each iteration
+        # Device = cpu/cuda
+        # GPU = gpu id
+        # rank = should be 0 for serial (dont change this)
+        # kwargs are for additional params (add as we go ahead)
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:521]: docstring

File: datasetinsights/estimators/siam_mask.py [hunk 20]
```diff
@@ -0,0 +1,2251 @@
...
+        overlap = IoU(
+            [x1, y1, x2, y2],
+            [target.x, target.y, target.x + target.w, target.y + target.h],
+        )
+        # Canonical BBox class does not support multiple IoUs together,
+        # so using the IoU from authors
+        # overlap = anyform2canonicalBBox([x1,y1,x2,y2],
+        # rep_type = "corner").iou(anyform2canonicalBBox
+        # (target,rep_type = "corner"))
+        # print("Overlap: ",overlap)
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]: comments

File: datasetinsights/estimators/siam_mask.py [hunk 21]
```diff
@@ -0,0 +1,2251 @@
...
+            name=attr,
+            val=float(self.val[attr]),
+            avg=float(self.sum[attr]) / self.count[attr],
+        )
+
+    def __getattr__(self, attr):
+        if attr in self.__dict__:
+            return super(AverageMeter, self).__getattr__(attr)
+        if attr not in self.sum:
+            print("invalid key '{}'".format(attr))
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:2246]: logging > print

File: datasetinsights/estimators/siam_mask.py [hunk 22]
```diff
@@ -0,0 +1,2251 @@
...
+        out = self.relu(out)
+
+        out = self.conv3(out)
+        out = self.bn3(out)
+
+        if self.downsample is not None:
+            residual = self.downsample(x)
+
+        if out.size() != residual.size():
+            print(out.size(), residual.size())
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:2005]: logging > print

File: datasetinsights/estimators/siam_mask.py [hunk 23]
```diff
@@ -0,0 +1,2251 @@
...
+
+        # log_once("p3 {}".format(p3.size()))
+        p4 = self.layer4(p3)
+
+        return p2, p3, p4
+
+
+# Bottleneck class - part of Resnet
+class Bottleneck(nn.Module):
+    expansion = 4
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:1958]: where should this live? is this a constant? a config hyperparameter? a property of the bottleneck itself?

File: datasetinsights/estimators/siam_mask.py [hunk 24]
```diff
@@ -0,0 +1,2251 @@
...
+                    self.unlock()
+                    return True
+                break
+        return False
+
+    # Set which layers will be trainable
+    def train_layers(self):
+        return self.layers[: self.train_num]
+
+    def unlock(self):
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:1817]: docstring should be clear about `unfix` vs `unlock` as the words are quite similar

File: datasetinsights/estimators/siam_mask.py [hunk 25]
```diff
@@ -0,0 +1,2242 @@
...
+    cx, cy = imw // 2, imh // 2
+    bbox = anyform2canonicalBBox([cx, cy, w, h], "center")
+    return bbox
+
+
+def crop_hwc(image, bbox, out_sz, padding=(0, 0, 0)):
+    """
+    Cropping image given a bounding box
+    Args: Image and modified bounding box
+    Returns: Cropped Image
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:109]: docstrings follow an indentation pattern: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: datasetinsights/estimators/siam_mask.py [hunk 26]
```diff
@@ -0,0 +1,2242 @@
...
+        borderValue=padding,
+    )
+    return crop
+
+
+def draw(image, box, name):
+    """
+        Args: image, bbox
+        Saves an image with bbox drawn on it
+        Only for debugging-> Not used anywhere
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:133]: docstring

File: datasetinsights/estimators/siam_mask.py [hunk 27]
```diff
@@ -0,0 +1,2242 @@
...
+    x1, y1, x2, y2 = int(box[0]), int(box[1]), int(box[2]), int(box[3])
+    image = image.copy()
+    # https://github.com/opencv/opencv/issues/14866#issuecomment-504632481 weird
+    cv2.rectangle(image, (x1, y1), (x2, y2), (0, 255, 0))
+    cv2.imwrite(name, image)
+
+
+"""
+    Generates a random number between 2 given numbers
+"""
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:149]: move into function.
format as docstring

File: datasetinsights/estimators/siam_mask.py [hunk 28]
```diff
@@ -0,0 +1,2242 @@
...
+    # (x) TODO: center crop both template and search -buggy
+    # make bounding box on both
+    # scale
+    # shift
+    # Any order below this
+    # grayscale
+    # blur
+    # resize - not tested
+    # flip - not tested
+    """
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]: docstring

File: datasetinsights/estimators/siam_mask.py [hunk 29]
```diff
@@ -0,0 +1,2242 @@
...
+    "center2corner": center2corner,
+    "random_number": genarate_random_number,
+}
+
+
+def get_transforms(center_crop, blur, gray, shift, scale, resize, flip):
+
+    """
+    # Should maintain this order
+    # (x) TODO: center crop both template and search -buggy
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]: what does this "buggy" bit mean?

File: datasetinsights/estimators/siam_mask.py [hunk 30]
```diff
@@ -0,0 +1,2242 @@
...
+            len(self.scales) * len(self.ratios) * (self.anchor_density ** 2)
+        )
+        self.anchors = None  # in single position (anchor_num*4)
+        self.all_anchors = None  # in all position 2*(4*anchor_num*h*w)
+        self.generate_anchors()
+
+    def generate_anchors(self):
+        """
+            Generates the anchor boxes with spec in config file
+        """
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py:306]: docstring

File: datasetinsights/estimators/siam_mask.py [hunk 31]
```diff
@@ -0,0 +1,2242 @@
...
+                        -h * 0.5 + y_offset,
+                        w * 0.5 + x_offset,
+                        h * 0.5 + y_offset,
+                    ][:]
+                    count += 1
+
+    def generate_all_anchors(self, im_c, size):
+        """
+            Helper function for generating anchors
+        """
```
Comment (zookae) [inline datasetinsights/estimators/siam_mask.py]: docstring

**Comment by adason** [comment]
Since dataseingisht will not store model training code, I am closing this issue to reduce confusion.
